File: src/main/java/org/cobbzilla/util/json/JsonEditOperation.java
package org.cobbzilla.util.json;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.databind.JsonNode;
import lombok.experimental.Accessors;

import java.io.IOException;
import java.util.List;

import static org.cobbzilla.util.json.JsonUtil.FULL_MAPPER;

@Accessors(chain=true)
public class JsonEditOperation {

    private JsonEditOperationType type;
    private String path;
    private String json;

    public boolean isRead() { return type == JsonEditOperationType.read; }

    public JsonNode getNode () throws IOException { return FULL_MAPPER.readTree(json); }

    public boolean hasIndex () { return getIndex() != null; }

    private List<String> initTokens() { return JsonUtil.tokenize(path); }
    @JsonIgnore
    private List<String> tokens = null;

    public boolean isEmptyBrackets () {
        int bracketPos = path.indexOf("[");
        int bracketClosePos = path.indexOf("]");
        return bracketPos != -1 && bracketClosePos != -1 && bracketClosePos == bracketPos+1;
    }

    public Integer getIndex() {
        List<String> tokens = getTokens();
        if (tokens.size() <= 1) return index(path);
        return index(tokens.get(tokens.size()-1));
    }

    private Integer index(String path) {
        try {
            int bracketPos = path.indexOf("[");
            int bracketClosePos = path.indexOf("]");
            if (bracketPos != -1 && bracketClosePos != -1 && bracketClosePos > bracketPos) {
                return new Integer(path.substring(bracketPos + 1, bracketClosePos));
            }
        } catch (Exception ignored) {}
        return null;
    }

    public String getName() {
        final List<String> tokens = getTokens();
        if (tokens.size() <= 1) return stripEmptyTrailingBrackets(path);
        return stripEmptyTrailingBrackets(tokens.get(tokens.size() - 1));
    }

    private String stripEmptyTrailingBrackets(String path) {
        return path.endsWith("[]") ? path.substring(0, path.length()-2) : path;
    }

    public int getNumPathSegments() { return getTokens().size(); }

    public String getName(int part) {
        return getTokens().get(part); }

    public JsonEditOperationType getType() {
        return this.type;
    }

    public String getPath() {
        return this.path;
    }

    public String getJson() {
        return this.json;
    }

    public List<String> getTokens() {
        if(tokens==null)tokens=initTokens();
        return this .tokens;
    }

    public JsonEditOperation setType(JsonEditOperationType type) {
        this.type = type;
        return this;
    }

    public JsonEditOperation setPath(String path) {
        this.path = path;
        return this;
    }

    public JsonEditOperation setJson(String json) {
        this.json = json;
        return this;
    }
}

File: src/main/java/org/cobbzilla/util/json/JsonEditOperation.kt
package org.cobbzilla.util.json

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.databind.JsonNode
import lombok.experimental.Accessors

import java.io.IOException

import org.cobbzilla.util.json.JsonUtil.FULL_MAPPER

@Accessors(chain = true)
class JsonEditOperation {

    private var type: JsonEditOperationType? = null
    private var path: String? = null
    private var json: String? = null

    val isRead: Boolean
        get() = type == JsonEditOperationType.read

    val node: JsonNode
        @Throws(IOException::class)
        get() = FULL_MAPPER.readTree(json)
    @JsonIgnore
    private var tokens: List<String>? = null

    val isEmptyBrackets: Boolean
        get() {
            val bracketPos = path!!.indexOf("[")
            val bracketClosePos = path!!.indexOf("]")
            return bracketPos != -1 && bracketClosePos != -1 && bracketClosePos == bracketPos + 1
        }

    val index: Int?
        get() {
            val tokens = getTokens()
            return if (tokens.size <= 1) index(path) else index(tokens[tokens.size - 1])
        }

    val name: String
        get() {
            val tokens = getTokens()
            return if (tokens.size <= 1) stripEmptyTrailingBrackets(path!!) else stripEmptyTrailingBrackets(tokens[tokens.size - 1])
        }

    val numPathSegments: Int
        get() = getTokens().size

    fun hasIndex(): Boolean {
        return index != null
    }

    private fun initTokens(): List<String> {
        return JsonUtil.tokenize(path!!)
    }

    private fun index(path: String?): Int? {
        try {
            val bracketPos = path!!.indexOf("[")
            val bracketClosePos = path.indexOf("]")
            if (bracketPos != -1 && bracketClosePos != -1 && bracketClosePos > bracketPos) {
                return Int(path.substring(bracketPos + 1, bracketClosePos))
            }
        } catch (ignored: Exception) {
        }

        return null
    }

    private fun stripEmptyTrailingBrackets(path: String): String {
        return if (path.endsWith("[]")) path.substring(0, path.length - 2) else path
    }

    fun getName(part: Int): String {
        return getTokens()[part]
    }

    fun getType(): JsonEditOperationType? {
        return this.type
    }

    fun getPath(): String? {
        return this.path
    }

    fun getJson(): String? {
        return this.json
    }

    fun getTokens(): List<String> {
        if (tokens == null) tokens = initTokens()
        return this.tokens
    }

    fun setType(type: JsonEditOperationType): JsonEditOperation {
        this.type = type
        return this
    }

    fun setPath(path: String): JsonEditOperation {
        this.path = path
        return this
    }

    fun setJson(json: String): JsonEditOperation {
        this.json = json
        return this
    }
}
